---
title: "Manuscript_Github_Pchabaudi_boilerplate"
author: "Timothy Little"
date: "29/07/2022"
output: html_document
---

Looking at GO enrichment of genes in- and out-of-phase between the 24 asexual life cycles of rMT and SBP transmitted P. chabaudi.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      results = TRUE, 
                      message = FALSE, 
                      warning = FALSE)

knitr::opts_knit$set(root.dir = 
                       ".")
#Note that the above doesn't give an error if the dir doesn't exist

options(scipen = 999) # Means that numbers are displayed normally not as 1e6 (1000000)
```

```{r loading_packages, include = FALSE, message = FALSE}
# Load the required packages

library('readr')
library('dplyr')
library('ggplot2')
library("RColorBrewer")
library('data.table')
library('tibble')
library('stringr')
library('kableExtra')
library('viridis')
library('reshape2')
library('ComplexHeatmap')
library('circlize') # for circular plots and colour
library("factoextra")
library("MatrixGenerics") # for rowMins and rowMax

#Colour blind friendly palette
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

#Want the defaults for arrange, mutate and summarise to be from dplyr
arrange <- dplyr::arrange
mutate <- dplyr::mutate
summarise <- dplyr::summarise
select <- dplyr::select

#Global save var - set to TRUE if you want to save figures.
global_save_var <- FALSE
```

```{r sourcing_functions}
source('../rfunctions/uniquify_fun.R')
source('../rfunctions/ordering_fun.R')
source('../rfunctions/tpm_fun.R')
source('../rfunctions/usefulFunctions_mround_my-max_firstup_fun.R')
source('../rfunctions/draw_heatmap_zscore.R')
```

```{r ordering_vec}
order_vector <- c("Early.Rings",
                  "Early.Mid.Rings",
                  "Late.Mid.Rings",
                  "Late.Rings",
                  "Rings.Troph.conversion",
                  "Early.Trophs",
                  "Mid.Trophs",
                  "Late.Trophs")
```

```{r cir_id, message=FALSE, results='hide'}
# The _cir_ gene names were downloaded using a search of PlasmoDB _P. chabaudi_ genome. _cir_ genes are the names of _pir_ genes in _P chabaudi_.
# Getting the cir gene names and transcript lengths needed for calculating the transcript-per-million

chabaudi_gene_info <- read.csv('../standard_files/PlasmoDB-47_beta_Pchabaudichabaudi_GenesByTaxon_Summary.csv',
                              header = TRUE,
                              stringsAsFactors = FALSE)

chabaudi_transcript_lengths <- select(chabaudi_gene_info, c(Gene.ID, Transcript.Length)) %>% 
  arrange(Gene.ID)

chabaudi_gene_info <- dplyr::rename(chabaudi_gene_info, 
                             Geneid = Gene.ID,
                             Description = Product.Description)

```

```{r cir_info}

cir_info <- read_delim('../standard_files/Pchabaudi_pir_info.txt',
           delim = '\t')
cir_id <- cir_info$Geneid

cir_clade_info <- read_delim('../standard_files/Pchabaudi_pir_clades.txt',
            delim = '\t')

#Add chapl_aapl/subfam numbers for easier heatmap drawing later.
cir_info <- cir_info %>% 
  group_by(chapl_aapl_name) %>% 
  mutate(chapl_aapl_name_num = n_distinct(Geneid)) %>% 
  mutate(
    chapl_aapl_name_num = paste0(
      chapl_aapl_name, 
      ' (n = ',
      chapl_aapl_name_num,
      ')'
    )) %>% 
  group_by(subfam) %>% 
  mutate(subfam_num = n_distinct(Geneid)) %>% 
  mutate(
    subfam_num = paste0(
      subfam, 
      ' (n = ',
      subfam_num,
      ')'
    )
  )
```

```{r times_stages_labels}
#Use the RMT stage differential count in order to name the samples, but keep time to retain order
times_stages_table <- data.frame(time = c('02h', '05h', '08h', '11h', '14h', '17h','20h','23h'),
                                 stage_diff_count = c('Late-Rings',
                                                      'Rings-Troph-conversion',
                                                      'Early-Trophs',
                                                      'Mid-Trophs',
                                                      'Late-Trophs',
                                                      'Early-Rings',
                                                      'Early-Mid-Rings',
                                                      'Late-Mid-Rings'))
# R will convert the dashes to '.' at some point so let's have this notation here also
times_stages_table$r_stage_diff_count <- str_replace_all(times_stages_table$stage_diff_count, 
                                                         pattern = '-', 
                                                         replacement = '\\.')
#Add a number from 1 to 8 for each stage, useful for phase calculations
times_stages_table$num <- 1:8
```

```{r colours_clades_phase}
SorL_col <- setNames(
  c(    '#F0E442',
        '#56B4E9', #L
        '#E69F00', #S
        '#999999'),
  sort(
    unique(cir_info$SorL)
  )
)

SorL_col_num <- SorL_col
names(SorL_col_num) <- sapply(names(SorL_col), 
                              function(clade_name) {
                                paste0(
                                  clade_name, 
                                  ' (n = ',
                                  sum(cir_info$SorL == clade_name),
                                  ')'
                                )
                              })

clade_col <- setNames(
  c('#F0E442',
    '#56B4E9', #L1
    '#009E73',
    '#0072B2',
    '#E69F00', #S1
    '#D55E00',
    '#CC79A7',
    '#999999'),
  sort(
    unique(cir_info$subfam)
  )
)

clade_col_num <- clade_col
names(clade_col_num) <- sapply(names(clade_col), 
                               function(clade_name) {
                                 paste0(
                                   clade_name, 
                                   ' (n = ',
                                   sum(cir_info$subfam == clade_name),
                                   ')'
                                 )
                               })

chapl_aapl_name_col <- setNames(
  c('#F0E442',
    '#56B4E9', #L1
    '#009E73',
    '#0072B2',
    '#E69F00', #S1
    '#D55E00',
    '#CC79A7',
    '#999999'),
  sort(
    unique(cir_info$chapl_aapl_name)
  )
)


chapl_aapl_name_col_num <- chapl_aapl_name_col
names(chapl_aapl_name_col_num) <- sapply(
  names(chapl_aapl_name_col), 
  function(clade_name) {
    paste0(
      clade_name, 
      ' (n = ',
      sum(cir_info$chapl_aapl_name == clade_name, 
          na.rm = TRUE),
      ')'
    )
  })
```

```{r read_TPM}
#Need to use the calculateTPM script first.

chab_tpm.df <- read_csv('data/Pchabaudi_TPM_all.csv')

chab_tpm_avg.df <- read_csv('data/Pchabaudi_TPM_average.csv')

chab_tpm_avg.list <- list(
  'RMT' = list(
    tpm = chab_tpm_avg.df[,grepl(colnames(chab_tpm_avg.df), 
                                pattern = 'rMT|Geneid')],
    experiment = 'RMT'
  ),
  'SBP' = list(
    tpm = chab_tpm_avg.df[,grepl(colnames(chab_tpm_avg.df), 
                                pattern = 'SBP|Geneid')],
    experiment = 'SBP'
  )
)

```

Adding in housekeeping genes for comparison:
HSP70 putative  PCHAS_0721000
MSP1 PCHAS_0831300
AMA1 PCHAS_0931000

```{r highest_expressed_stage_heatmap_fig3b}

housekeeping_genes <- data.frame('HSP70' = 'PCHAS_0721000',
                                 'MSP1' = 'PCHAS_0831300',
                                 'AMA1' = 'PCHAS_0931000')

cir_housekeeping_id <- c(cir_id, 
                         'PCHAS_0721000',
                         'PCHAS_0831300',
                         'PCHAS_0931000')

# Get order of pir genes by highest expressed genes in each sample.

combined_data_pir <-  filter(chab_tpm_avg.df,
                             Geneid %in% cir_housekeeping_id) %>% 
  column_to_rownames(var = 'Geneid') %>% 
  filter_all(
    any_vars(. > 10)
  )

```

```{r}
#Ordering that a more understandable order

combined_data_ordered <- tibble(Geneid = names(apply(combined_data_pir, 1, mean)),
                                means = apply(combined_data_pir, 1, mean),
                                subfam = factor(
                                  ifelse(rownames(combined_data_pir) %in% cir_id,
                                         cir_info$subfam[match(rownames(combined_data_pir), 
                                                               cir_info$Geneid)],
                                         'ref')
                                )) %>% 
  mutate(subfam =
           forcats::fct_shift(
             relevel(
               as.factor(subfam),
               ref = 'ref')
           )
  ) %>% 
  # filter(means > 1) %>% 
  arrange(subfam, 
          desc(means)) %>% 
  .$Geneid

#Order the heatmap data by the order defined above (Subfam, then mean expression).
combined_data_pir_ht <- combined_data_pir[
  order(
    match(rownames(combined_data_pir),
          combined_data_ordered),
    na.last = NA #Remove NAs
  ),
]

#Change the colnames to the times
colnames(combined_data_pir_ht) <- str_extract(colnames(combined_data_pir_ht),
                                              pattern = '.{3}_([[:digit:]]{2}h|sch.culture)')

#Order by time point
combined_data_pir_ht <- combined_data_pir_ht[,order(colnames(combined_data_pir_ht)),
                                             drop = FALSE] 
```

```{r}
#Add a column for the subfam (ref for the reference genes)
combined_data_pir_ht$subfam <- ifelse(rownames(combined_data_pir_ht) %in% cir_id,
                                      cir_info$subfam_num[match(rownames(combined_data_pir_ht), 
                                                            cir_info$Geneid)],
                                      'ref')

#Get the ref gene actual names instead of the Geneid
rownames(combined_data_pir_ht) <- ifelse(
  rownames(combined_data_pir_ht) %in% housekeeping_genes[1,],
  colnames(housekeeping_genes)[ match(rownames(combined_data_pir_ht), housekeeping_genes[1,]) ],
  rownames(combined_data_pir_ht)
)

#Get the colour range
maxcol <- ceiling(
  max(
    log2(1+combined_data_pir_ht[!colnames(combined_data_pir_ht) %in% 'subfam'])
  ))

#Sub-family annotation
clade_info <- combined_data_pir_ht$subfam
annotation_col <- c(clade_col, 'ref' = 'black')

#Subfamily annotation

clade_ha <- rowAnnotation(
  'sub-family' = anno_simple(
    clade_info,
    col = annotation_col,
  ),
  annotation_name_gp = gpar(fontsize = 10)
)

#Subfamily name labels
num_subfam <- cir_info %>% 
  group_by(subfam) %>% 
  summarise(num = n_distinct(Gene.ID))

labels_subfam <- c(setNames(paste0(num_subfam$subfam, ' (n = ', num_subfam$num, ')'),
                            num_subfam$subfam),
                   'ref' = 'ref')
```

```{r draw_heatmap}
ht.list <- lapply(c('rMT', 'SBP'),
                  function(transmission){
                    
                    ht.df <- combined_data_pir_ht[
                      ,
                      grepl(colnames(combined_data_pir_ht), 
                            pattern = paste0(transmission))
                    ]
                    
                    colnames(ht.df) <- str_remove(colnames(ht.df),
                                                  pattern = paste0(transmission, 
                                                                   '_'))
                    
                    ht <- Heatmap(
                      log2(
                        1+ht.df
                      ),
                      cluster_rows = FALSE,
                      cluster_columns = FALSE,
                      show_row_names = FALSE,
                      name = 'log(TPM+1)', 
                      col = colorRamp2(c(0,
                                         maxcol/4,
                                         maxcol/2,
                                         3*maxcol/4,
                                         maxcol),
                                       viridis(5)),
                      na_col = 'white', 
                      border = TRUE,
                      border_gp = gpar(lwd = 0.1),
                      column_names_gp = gpar(fontsize = 10),
                      row_dend_width = unit(3, "cm"),
                      width = unit(4.5, 'cm'),
                      height = unit(7,'cm'),
                      column_title = paste(transmission),
                      row_title_rot = 0,
                      row_title_gp = gpar(fontsize = 10),
                      row_split = forcats::fct_shift(
                        relevel(
                          as.factor(combined_data_pir_ht$subfam),
                          ref = 'ref')
                      ),
                      heatmap_legend_param = list(
                        labels_gp = gpar(fontsize = 10),
                        title_gp = gpar(fontsize = 10,
                                        fontface = 'bold'),
                        at = c(0,
                               maxcol/4,
                               maxcol/2,
                               3*maxcol/4,
                               maxcol)
                      )
                    )
                    return(ht)
                  })

ht.list <- Reduce('+', ht.list)

clade_lgd <- Legend(
  title = 'sub-family',
  at = c(labels_subfam),
  legend_gp = gpar(fill = annotation_col)
)

if(global_save_var){
  pdf(file = 'plots/subfam_TPMheatmap.pdf',
      height = 5,
      width = 8)
  draw(ht.list, 
       row_title_side = 'left',
       column_title_side = 'bottom',
       heatmap_legend_side = 'right',
       merge_legend = TRUE)
  dev.off()
}
```
