---
title: "Manuscript_Github_Pchabaudi_phaseGene"
author: "Timothy Little"
date: "27/07/2022"
output: html_document
---

Looking at GO enrichment of genes in- and out-of-phase between the 24 asexual life cycles of rMT and SBP transmitted P. chabaudi.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      results = TRUE, 
                      message = FALSE, 
                      warning = FALSE)

knitr::opts_knit$set(root.dir = 
                       ".")
#Note that the above doesn't give an error if the dir doesn't exist

options(scipen = 999) # Means that numbers are displayed normally not as 1e6 (1000000)
```

```{r loading_packages, include = FALSE, message = FALSE}
# Load the required packages

library('readr')
library('dplyr')
library('ggplot2')
library("RColorBrewer")
library('data.table')
library('tibble')
library('stringr')
library('kableExtra')
library('viridis')
library('reshape2')
library('ComplexHeatmap')
library('circlize') # for circular plots and colour
library("factoextra")
library("MatrixGenerics") # for rowMins and rowMax
library("topGO")
library("FoldGO")
library('Rgraphviz')

#Colour blind friendly palette
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

#Want the defaults for arrange, mutate and summarise to be from dplyr
arrange <- dplyr::arrange
mutate <- dplyr::mutate
summarise <- dplyr::summarise
select <- dplyr::select

#Global save var - set to TRUE if you want to save figures.
global_save_var <- FALSE
```

```{r sourcing_functions}
source('../rfunctions/uniquify_fun.R')
source('../rfunctions/ordering_fun.R')
source('../rfunctions/tpm_fun.R')
source('../rfunctions/usefulFunctions_mround_my-max_firstup_fun.R')
source('../rfunctions/draw_heatmap_zscore.R')
```

```{r ordering_vec}
order_vector <- c("Early.Rings",
                  "Early.Mid.Rings",
                  "Late.Mid.Rings",
                  "Late.Rings",
                  "Rings.Troph.conversion",
                  "Early.Trophs",
                  "Mid.Trophs",
                  "Late.Trophs")
```

```{r cir_id, message=FALSE, results='hide'}
# The _cir_ gene names were downloaded using a search of PlasmoDB _P. chabaudi_ genome. _cir_ genes are the names of _pir_ genes in _P chabaudi_.
# Getting the cir gene names and transcript lengths needed for calculating the transcript-per-million

chabaudi_gene_info <- read.csv('../standard_files/PlasmoDB-47_beta_Pchabaudichabaudi_GenesByTaxon_Summary.csv',
                              header = TRUE,
                              stringsAsFactors = FALSE)

chabaudi_transcript_lengths <- select(chabaudi_gene_info, c(Gene.ID, Transcript.Length)) %>% 
  arrange(Gene.ID)

chabaudi_gene_info <- dplyr::rename(chabaudi_gene_info, 
                             Geneid = Gene.ID,
                             Description = Product.Description)

```

```{r cir_info}

cir_info <- read_delim('../standard_files/Pchabaudi_pir_info.txt',
           delim = '\t')
cir_id <- cir_info$Geneid

cir_clade_info <- read_delim('../standard_files/Pchabaudi_pir_clades.txt',
            delim = '\t')
```

```{r times_stages_labels}
#Use the RMT stage differential count in order to name the samples, but keep time to retain order
times_stages_table <- data.frame(time = c('02h', '05h', '08h', '11h', '14h', '17h','20h','23h'),
                                 stage_diff_count = c('Late-Rings',
                                                      'Rings-Troph-conversion',
                                                      'Early-Trophs',
                                                      'Mid-Trophs',
                                                      'Late-Trophs',
                                                      'Early-Rings',
                                                      'Early-Mid-Rings',
                                                      'Late-Mid-Rings'))
# R will convert the dashes to '.' at some point so let's have this notation here also
times_stages_table$r_stage_diff_count <- str_replace_all(times_stages_table$stage_diff_count, 
                                                         pattern = '-', 
                                                         replacement = '\\.')
#Add a number from 1 to 8 for each stage, useful for phase calculations
times_stages_table$num <- 1:8
```

```{r read_TPM}
#Need to use the calculateTPM script first.

chab_tpm.df <- read_csv('data/Pchabaudi_TPM_all.csv')

chab_tpm_avg.df <- read_csv('data/Pchabaudi_TPM_average.csv')

chab_tpm_avg.list <- list(
  'RMT' = list(
    tpm = chab_tpm_avg.df[,grepl(colnames(chab_tpm_avg.df), 
                                pattern = 'rMT|Geneid')],
    experiment = 'RMT'
  ),
  'SBP' = list(
    tpm = chab_tpm_avg.df[,grepl(colnames(chab_tpm_avg.df), 
                                pattern = 'SBP|Geneid')],
    experiment = 'SBP'
  )
)

```

```{r import_z}
chab_z.df <- read.csv(file = 'data/Pchabaudi_z.csv')

#Make list object of the TPM values and the z-scores value by concatenating the 
# lists of the zscore dataframes and the TPM lists from above.
chab_z.list <- list('RMT' = c(chab_tpm_avg.list[['RMT']],
                              list('zscores' = chab_z.df[,grepl(colnames(chab_z.df),
                                                           pattern = 'Geneid|rMT')])),
                    'SBP' = c(chab_tpm_avg.list[['SBP']],
                              list('zscores' = chab_z.df[,grepl(colnames(chab_z.df),
                                                           pattern = 'Geneid|SBP')])))

order_z_rMT <- read.table(file = 'data/Pchabaudi_z_phaseOrder_rMT.list')[,1]
order_z_SBP <- read.table(file = 'data/Pchabaudi_z_phaseOrder_SBP.list')[,1]

```

```{r colours_clades_phase}
SorL_col <- setNames(
  c(    '#F0E442',
        '#56B4E9', #L
        '#E69F00', #S
        '#999999'),
  sort(
    unique(cir_info$SorL)
  )
)

SorL_col_num <- SorL_col
names(SorL_col_num) <- sapply(names(SorL_col), 
                              function(clade_name) {
                                paste0(
                                  clade_name, 
                                  ' (n = ',
                                  sum(cir_info$SorL == clade_name),
                                  ')'
                                )
                              })

clade_col <- setNames(
  c('#F0E442',
    '#56B4E9', #L1
    '#009E73',
    '#0072B2',
    '#E69F00', #S1
    '#D55E00',
    '#CC79A7',
    '#999999'),
  sort(
    unique(cir_info$subfam)
  )
)

clade_col_num <- clade_col
names(clade_col_num) <- sapply(names(clade_col), 
                               function(clade_name) {
                                 paste0(
                                   clade_name, 
                                   ' (n = ',
                                   sum(cir_info$subfam == clade_name),
                                   ')'
                                 )
                               })


```

Working out whether a given gene is out-of-sync based on when it peaks in transcription. If it peaks at different timepoints in rMT compared to SBP, then the time difference between these points is considered the gene's synchronicity. 0h refers to a gene that's transcriped maximally at the same time in rMT and SBP infections. 

Note that this isn't directional; 6h out-of-sync could be 6h before or after.

# Number of genes out-of-phase.

```{r gene_sync}

#Which genes are out of sync between rMT and SBP?

#Melt the z-score dataframe into long form, then slice out the maximum z-score value.
# Filter out any gene which has multiple max values for simplicity.
# Summarise this with the numbers that each stage (the max stage) corresponds to.

z_max <- lapply(chab_z.list, 
                function(x) {
  x$zscores %>% 
  reshape2::melt(id.vars = 'Geneid', 
       variable.name = 'sample', 
       value.name = 'zscore') %>% 
  group_by(Geneid) %>% 
  slice_max(order_by = zscore, n = 1) %>% 
  filter(n() == 1) %>% 
  summarise(sample_max = times_stages_table$num[
    str_detect(sample, 
               times_stages_table$time)
    ])
})

#Bind the two together with inner_join so that it only includes genes in both.
# Then calculate the peak difference by subtracting the difference between the sample numbers.
# Since this is a cycle, no point of the cycle is theoretically 
#  more than 4 time points away from the next one. 
#  e.g. 23h and 02h (8 and 1) are only 1 time point apart, not seven.
#  Hence we subtract values above four from eight
z_max_diff <- inner_join(z_max[[1]], z_max[[2]], by = 'Geneid') %>% 
  mutate(diff = ifelse(abs(sample_max.x - sample_max.y) > 4,
                       8 - abs(sample_max.x - sample_max.y),
                       abs(sample_max.x - sample_max.y)))

#Probably unnecessarily complicated code to make a barplot of how many genes 
# are peaking at different timepoints between rMT and SBP
barplot(group_by(z_max_diff, diff) %>% summarise(total = n()) %>% .$total,
        names.arg = paste0(sort(unique(z_max_diff$diff))*3, 'h'),
        main = 'all genes out-of-phase')

#And for the pir genes
barplot(group_by(z_max_diff, diff) %>% 
          filter(Geneid %in% cir_id) %>% 
          summarise(total = n()) %>% 
          .$total,
        names.arg = paste0(sort(unique(z_max_diff$diff))*3, 'h'),
        main = 'pir genes out-of-phase')

```

```{r z_genelist}
#Get the gene list for each group of genes in and out of phase.
gene.list <- chabaudi_gene_info$Geneid
z_max_diff.list <- lapply(seq(0,4), 
                          function(diff_num) {
                            diff.list <- filter(z_max_diff, diff == diff_num) %>% 
                              .$Geneid
                            #Get 1 for on the list and 0 for not.
                            gene_diff.list <- as.integer(gene.list %in% diff.list)
                            names(gene_diff.list) <- gene.list
                            return(gene_diff.list)
                          })
names(z_max_diff.list) <- paste0(seq(0,4)*3, 'h')

```

```{r phaseograms}
lapply(paste0(seq(0,4)*3, 'h'), 
       function(phase){
         draw_heatmap_zscore(zscore_list = chab_z.list,
                             filter_genes_vector = names(
                               getElement(z_max_diff.list, paste(phase))[
                                 getElement(z_max_diff.list, paste(phase)) == 1
                               ]
                             ),
                             gene_ordering_vector = order_z_rMT,
                             threshold_tpm = 1,
                             use_times = TRUE,
                             use_order_vector = FALSE,
                             show_row_names = FALSE,
                             tpm_mat = TRUE,
                             column_title_name = paste0('Genes ',phase, ' out-of-phase'))
         return()
       })
```

The out-of-phase values represent the data reasonably well, although it is notable that for some genes only 3 or 6h out-of-sync the peak of transcription is quite broad and overlaps.

# How many of these genes do not have GO annotations?

```{r phase_go_enrich}

#Read in the GO-Annotation file using GAFReader from FoldGO.
# getAnnotation to obtain a list of the GO terms and associated genes.
# inverseList (from topGO) to turn this into list of geneID with associated GO terms.
gaf <- GAFReader("../standard_files/PlasmoDB-48_Pchabaudichabaudi_GO.gaf")
go_annotation <- inverseList(getAnnotation(gaf))

# How many genes for each group do not have any GO annotations?

lapply(z_max_diff.list,
       function(gene_diff.list){
         total_noGO <- sum(!names(gene_diff.list[gene_diff.list == 1]) %in% 
                             names(go_annotation))
         prop_noGO <- total_noGO/sum(gene_diff.list == 1)
         return(paste0(total_noGO, ' genes without annotation = ',
                       round(prop_noGO*100, digits = 2), '%'))
       })

```

# GO term enrichment.

Perform a Fishers test using values of 1 for genes of out-of-sync group being tested, and 0 for all others.

## Cellular Component.

```{r go-list}
# Use the topGO functions to make a topGOdata object, 
# conduct a Fisher test for statistical significance, 
# then create a table of these signifcant GO terms.
GOres.list <- lapply(paste0(seq(0,4)*3, 'h'), 
                     function(phase){
                       GOdata <- new("topGOdata", 
                                     ontology = "CC", 
                                     geneSel = function(x){x == 1},
                                     allGenes = getElement(z_max_diff.list, paste(phase)),
                                     nodeSize = 5,
                                     annot = annFUN.gene2GO, 
                                     gene2GO = go_annotation)
                       resultFis <- runTest(GOdata, 
                                            algorithm = "classic", 
                                            statistic = "fisher")
                       gentable <- GenTable(GOdata, 
                                              classic = resultFis, 
                                              orderBy = "classic", 
                                              ranksOf = "classic",
                                              topNodes = 25) %>% 
                           filter(classic < 0.05)
                       return(list(GOdata = GOdata,
                                   resultFis = resultFis,
                                   gentable = gentable))
                     })
names(GOres.list) <- paste0(seq(0,4)*3, 'h')
lapply(GOres.list, getElement, 'gentable')

```

Phaseograms of the genes represented by the most significant GO term returned for each out-of-sync group. 

These phaseograms show every gene in the entire dataset associated with the 1st or 2nd most statistically signif enriched GO term, so they may not all be as out-of-phase as one another. It is likely that only a subset of GO term genes will be out-of-sync by the same number of timepoints but this can be sufficient still for the GO term to be enriched. See the tables above for the number of genes in that GO term group ('Annotated') and the number that are in that out-of-sync group ('Significant').

This is a good test to see if this GO enrichment seems correct - if most of the genes look like they're not in the phase expected then the GO enrichment is probably erroneous.

```{r topGO_phase}

lapply(names(GOres.list), 
       function(GOres.num){
         GOres <- getElement(GOres.list, GOres.num)
         lapply(c(1,2), 
                function(which_go){
                  #Get first or second (most significant) enriched GO id.
                  goID <- GOres$gentable[which_go, "GO.ID"]
                  goTerm <- GOres$gentable[which_go, "Term"]
                  
                  #Show distribution of GO term genes compared to null distribution (all other genes)
                  # print(showGroupDensity(GOres$GOdata, goID, ranks = TRUE))
                  
                  #Can use genesInTerm function to extract the genes related to that GO term
                  genesinterm <- genesInTerm(GOres$GOdata, goID)
                  goID_genes <- unlist(genesinterm)
                  names(goID_genes) <- genesinterm
                  
                  draw_heatmap_zscore(zscore_list = chab_z.list,
                                      filter_genes_vector = goID_genes,
                             gene_ordering_vector = order_z_rMT,
                             threshold_tpm = 1,
                             use_times = TRUE,
                             use_order_vector = FALSE,
                             show_row_names = FALSE,
                             tpm_mat = TRUE,
                             column_title_name = paste0(GOres.num, ' - ',
                                                        which_go, ' most sig',
                                                        ' - ', goTerm)
         )
         return()
                })
         return()
       })

```

Note that some enriched GO terms may not appear in the annotation file (.gaf) directly as they are found to be enriched through the network of related/ancestral terms from the GO network of relations.

## Biological Process

```{r go-list-bp}
# Use the topGO functions to make a topGOdata object, 
# conduct a Fisher test for statistical significance, 
# then create a table of these signifcant GO terms.
GOres.bp.list <- lapply(paste0(seq(0,4)*3, 'h'), 
                     function(phase){
                       GOdata <- new("topGOdata", 
                                     ontology = "BP", 
                                     geneSel = function(x){x == 1},
                                     allGenes = getElement(z_max_diff.list, paste(phase)),
                                     nodeSize = 5,
                                     annot = annFUN.gene2GO, 
                                     gene2GO = go_annotation)
                       resultFis <- runTest(GOdata, 
                                            algorithm = "classic", 
                                            statistic = "fisher")
                       gentable <- GenTable(GOdata, 
                                              classic = resultFis, 
                                              orderBy = "classic", 
                                              ranksOf = "classic",
                                              topNodes = 25) %>% 
                           filter(classic < 0.05)
                       return(list(GOdata = GOdata,
                                   resultFis = resultFis,
                                   gentable = gentable))
                     })
names(GOres.bp.list) <- paste0(seq(0,4)*3, 'h')
lapply(GOres.bp.list, getElement, 'gentable')

```


```{r topGO_phase-bp}

lapply(names(GOres.bp.list), 
       function(GOres.num){
         GOres <- getElement(GOres.bp.list, GOres.num)
         lapply(c(1,2), 
                function(which_go){
                  #Get first or second (most significant) enriched GO id.
                  goID <- GOres$gentable[which_go, "GO.ID"]
                  goTerm <- GOres$gentable[which_go, "Term"]
                  
                  #Show distribution of GO term genes compared to null distribution (all other genes)
                  # print(showGroupDensity(GOres$GOdata, goID, ranks = TRUE))
                  
                  #Can use genesInTerm function to extract the genes related to that GO term
                  genesinterm <- genesInTerm(GOres$GOdata, goID)
                  goID_genes <- unlist(genesinterm)
                  names(goID_genes) <- genesinterm
                  
                  draw_heatmap_zscore(zscore_list = chab_z.list,
                                      filter_genes_vector = goID_genes,
                             gene_ordering_vector = order_z_rMT,
                             threshold_tpm = 1,
                             use_times = TRUE,
                             use_order_vector = FALSE,
                             show_row_names = FALSE,
                             tpm_mat = TRUE,
                             column_title_name = paste0(GOres.num, ' - ',
                                                        which_go, ' most sig',
                                                        ' - ', goTerm)
         )
         return()
                })
         return()
       })

```

Noticeable enrichments of genes that are:

* In-phase: DNA and nucleotide processes, and amino acid processing, such as DNA replication, tRNA aminoacylation and amino acid activation.
* 3h Out-of-phase: Translation and peptide biosynthesis, gene expression and ribosome biogenesis.
* 6h Out-of-phase: (Based on much fewer genes) Metabolism of carbohydrates and respiration.
* 9h Out-of-phase: (Based on even fewerer genes - phaseogram not quite convincing) More metabolism and lipid biosynthesis.
* 12h Out-of-phase: (Based on a few more genes actually - phaseogram not quite convincing) mRNA processing.