---
title: "Manuscript_Github_Pchabaudi_calculateTPM"
author: "Timothy Little"
date: "22/03/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      results = TRUE, 
                      message = FALSE, 
                      warning = FALSE)

knitr::opts_knit$set(root.dir = 
                       ".")
#Note that the above doesn't give an error if the dir doesn't exist

options(scipen = 999) # Means that numbers are displayed normally not as 1e6 (1000000)
```

```{r loading_packages, include = FALSE, message = FALSE}
# Load the required packages

library('readr')
library('dplyr')
library('ggplot2')
# library("pheatmap")
library("RColorBrewer")
library('data.table')
library('gridExtra')
library('tibble')
library('stringr')
library('kableExtra')
library('viridis')
# library('corrplot')
# library('reshape2')
# library('circlize') # for circular plots and colours
# library('cluster')    # clustering algorithms
# library('factoextra') # clustering algorithms & visualization
# library('ggfortify') # so ggplot understands prcomp objects
# library('ape') # for phylogeny plotting
# library('dendextend') # for dendrogram plotting
# library('readxl') # for reading in Excel spreadsheets
# library('plotly')
# library('scales')
# library('processx') # for downloading the plotly plots
# library('cowplot')
# library('ComplexHeatmap')
# library('ggpattern')

#Colour blind friendly palette
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
#show_col(cbPalette)

#Want the defaults for arrange, mutate and summarise to be from dplyr
arrange <- dplyr::arrange
mutate <- dplyr::mutate
summarise <- dplyr::summarise
select <- dplyr::select
```

```{r sourcing_functions}
source('../rfunctions/uniquify_fun.R')
source('../rfunctions/ordering_fun.R')
source('../rfunctions/tpm_fun.R')
source('../rfunctions/usefulFunctions_mround_my-max_firstup_fun.R')
```

```{r order_vec}

order_vector <- c("Early.Rings",
                  "Early.Mid.Rings",
                  "Late.Mid.Rings",
                  "Late.Rings",
                  "Rings.Troph.conversion",
                  "Early.Trophs",
                  "Mid.Trophs",
                  "Late.Trophs")
```

The _cir_ gene names were downloaded using a search of PlasmoDB _P. chabaudi_ genome. _cir_ genes are the names of _pir_ genes in _P chabaudi_.

```{r cir_id, message=FALSE, results='hide'}

# Getting the cir gene names and transcript lengths needed for calculating the transcript-per-million

chabaudi_gene_info <- read.csv('../standard_files/PlasmoDB-47_beta_Pchabaudichabaudi_GenesByTaxon_Summary.csv',
                              header = TRUE,
                              stringsAsFactors = FALSE)

chabaudi_transcript_lengths <- select(chabaudi_gene_info, c(Gene.ID, Transcript.Length)) %>% 
  arrange(Gene.ID) %>% 
  mutate(Geneid = Gene.ID) %>% 
  select(Geneid, Transcript.Length)

chabaudi_gene_info <- dplyr::rename(chabaudi_gene_info, 
                             Geneid = Gene.ID,
                             Description = Product.Description)

```

```{r cir_info}

cir_info <- read_delim('../standard_files/Pchabaudi_pir_info.txt',
           delim = '\t')
cir_id <- cir_info$Geneid

cir_clade_info <- read_delim('../standard_files/Pchabaudi_pir_clades.txt',
            delim = '\t')
```

```{r times_stages_labels}
#Use the rMT stage differential count in order to name the samples, but keep time to retain order
times_stages_table <- data.frame(time = c('02h', '05h', '08h', '11h', '14h', '17h','20h','23h'),
                                 stage_diff_count = c('Late-Rings',
                                                      'Rings-Troph-conversion',
                                                      'Early-Trophs',
                                                      'Mid-Trophs',
                                                      'Late-Trophs',
                                                      'Early-Rings',
                                                      'Early-Mid-Rings',
                                                      'Late-Mid-Rings'))
# R will convert the dashes to '.' at some point so let's have this notation here also
times_stages_table$r_stage_diff_count <- str_replace_all(times_stages_table$stage_diff_count, 
                                                         pattern = '-', 
                                                         replacement = '\\.')
```

```{r calc_tpm}
#Read in the count data and remove one sample which is unknown and had no description.
chab_count_table <- read.csv('data/tl19-07_twentyfour_counts_samples.csv') %>% 
  select(-rMT_Unknown)
#Calculate the TPM
chab_tpm.df <- tpm(count_data = chab_count_table, 
                          lengths = chabaudi_transcript_lengths)

#Do they all sum to 1e6? Note that all.equal allows for a tiny bit of difference which is okay.
all.equal(apply(chab_tpm.df[-1], 2, sum), 
          rep(1e6, times = dim(chab_tpm.df[-1])[2]), 
          check.names = FALSE)
```

```{r check_tpm}

#Used `dput` function to create a df of randomly chosen part of what I assume to be the true values.

check_tpm <- structure(list(Geneid = c("PCHAS_1236900", "PCHAS_1237000", "PCHAS_1237100", 
"PCHAS_1237200", "PCHAS_1237300", "PCHAS_1237400"), rMT_17h_Early.Rings_1 = c(9.76166246727995, 
2.53080138040591, 10.9126974974354, 197.447300616447, 27.0792042284135, 
55.9778949394867), rMT_sch.culture_1 = c(4.82209836798507, 39.9498029457425, 
4.36161304058411, 192.21404227623, 255.05821864658, 252.312788862482
), rMT_20h_Early.Mid.Rings_1 = c(17.0998539251424, 7.33599453376897, 
12.8493972283173, 213.998967387733, 11.0784616329143, 60.1572274917456
), rMT_05h_Rings.Troph.conversion_1 = c(38.3647026622499, 6.90734779639621, 
2.62130156899195, 194.546775103218, 3.78235443023746, 130.068871216885
), rMT_sch.culture_2 = c(2.90683176639546, 22.597581594288, 4.18468334820306, 
368.645503106066, 438.239482969298, 330.318886377763), rMT_05h_Rings.Troph.conversion_2 = c(32.6510468524069, 
7.16075860604918, 1.31010772565867, 164.370211763169, 2.96069000666786, 
109.168242066326), rMT_11h_Mid.Trophs_1 = c(26.147595680817, 
27.0999659114182, 2.16752322103163, 541.644380621737, 2.53144630329207, 
120.735400671386), rMT_14h_Late.Trophs_1 = c(11.4079837593492, 
22.856112916092, 10.5413074421478, 951.44994461046, 73.4132553701819, 
131.681439654117), rMT_02h_Late.Rings_1 = c(31.2811649503316, 
6.57534976948559, 3.85345590947847, 150.532386139873, 3.9545205115986, 
94.4792421254667), rMT_08h_Early.Trophs_1 = c(20.157812092634, 
12.1280059532542, 0.650734564771486, 170.309456869319, 1.59767139279832, 
83.2278478943328), rMT_17h_Early.Rings_2 = c(17.403862005927, 
3.69172830428755, 20.6004897213014, 250.122760155978, 17.2424791957208, 
59.8810845288676), rMT_23h_Late.Mid.Rings_1 = c(29.5379734077001, 
5.22489884006776, 3.67411111636422, 170.523646451821, 4.16641042383719, 
82.6825661834616), rMT_23h_Late.Mid.Rings_2 = c(32.6253792704758, 
5.07505899762957, 3.70087210432098, 163.457320547547, 2.80131367804736, 
95.4154100528915), rMT_20h_Early.Mid.Rings_2 = c(9.98436534398298, 
8.86560981918151, 15.9408661685613, 168.519812840241, 18.4953322269757, 
67.1081583940078), rMT_11h_Mid.Trophs_2 = c(16.8514270116997, 
15.0710551161147, 1.34774236993737, 283.638146433073, 2.88781264078723, 
104.469069579326), rMT_17h_Early.Rings_3 = c(9.80685483532316, 
7.8009072553707, 12.9140266989638, 244.004094798698, 32.6454833873949, 
61.5795685612941), rMT_05h_Rings.Troph.conversion_3 = c(44.6914009986881, 
7.31985380744853, 2.09226783944416, 202.848558910157, 3.38742465612069, 
108.563983207627), rMT_14h_Late.Trophs_2 = c(16.1568208764581, 
25.859815232017, 5.9401546141338, 849.643405800935, 7.56694623416709, 
138.133764925281), rMT_08h_Early.Trophs_2 = c(27.5234173956934, 
13.6584098001981, 2.54054386371032, 341.123228542418, 2.10140240005804, 
132.057622012364), rMT_20h_Early.Mid.Rings_3 = c(20.2528241267272, 
5.33303205844855, 11.6247077258001, 252.801844519901, 11.208737218013, 
96.11509755057), rMT_sch.culture_3 = c(7.02704565010117, 31.941116591369, 
10.1000976836731, 459.609650763128, 291.376205366797, 211.18569154735
), rMT_02h_Late.Rings_2 = c(41.5030181049848, 8.57199496745699, 
6.57168269134606, 260.300795278229, 8.65645647459549, 115.867220110965
), rMT_11h_Mid.Trophs_3 = c(23.668981121663, 17.3111675994721, 
2.40864504534792, 403.270229427547, 3.31051807265358, 107.600469742135
), SBP_02h_Late.Rings_1 = c(32.1039867447089, 4.18890973084741, 
1.83235045534882, 253.342646772667, 3.24971327904687, 147.599200658419
), SBP_05h_Rings.Troph.conversion_1 = c(69.9933903732369, 5.56765605241657, 
3.51252089834496, 253.482911605055, 3.94202779803947, 149.365010977175
), SBP_08h_Early.Trophs_1 = c(67.4596701130964, 8.04708690737168, 
1.29888044701547, 280.113721889105, 2.32961719999114, 168.821435727764
), SBP_11h_Mid.Trophs_1 = c(39.0509461034118, 20.3431512112609, 
2.45091379976465, 870.791722784726, 2.13865338832938, 218.047682451552
), SBP_14h_Late.Trophs_1 = c(31.0258753755895, 24.8362618680581, 
8.02873672984643, 1248.50319033523, 8.08944053716688, 165.499381325518
), SBP_17h_Early.Rings_1 = c(21.5890368368948, 23.7391769164685, 
7.41687066698247, 827.388691990966, 97.5224751506972, 187.900653468842
), SBP_20h_Early.Mid.Rings_1 = c(4.91288997903853, 6.15026465334183, 
7.00690865862872, 183.461124643034, 40.6941010432805, 94.2171898730868
), SBP_23h_Late.Mid.Rings_1 = c(17.0960581250041, 4.46989348422346, 
2.59223414646292, 226.408675568144, 8.83071196883189, 130.839085546676
), SBP_02h_Late.Rings_2 = c(36.1370878333684, 2.84224286329863, 
3.76479788403891, 193.540809015887, 4.86746227785013, 102.682045137645
), SBP_05h_Rings.Troph.conversion_2 = c(55.9703217173144, 5.09338376368401, 
1.6164154903665, 326.007322410062, 2.84115197098826, 154.530538517825
), SBP_08h_Early.Trophs_2 = c(59.1292861603963, 6.89801752788374, 
1.62611489888877, 302.374971383912, 2.28208571581952, 159.285259970324
), SBP_05h_Rings.Troph.conversion_3 = c(64.0277646953059, 4.36881423377574, 
5.38898632230727, 265.098915160974, 6.74434224459245, 115.619635926338
), SBP_11h_Mid.Trophs_2 = c(56.2225636049175, 14.0211061570003, 
1.53632625772768, 440.630839066675, 2.67215322318625, 180.373213104462
), SBP_14h_Late.Trophs_2 = c(30.676698797827, 27.1931566057931, 
6.91186772412193, 1614.99745495943, 22.3622898426203, 156.841577995753
), SBP_17h_Early.Rings_2 = c(14.7757154238053, 15.693441999666, 
10.0058832290506, 688.261040899689, 103.363777591419, 166.096400179785
), SBP_20h_Early.Mid.Rings_2 = c(7.1021602822451, 4.3088000205048, 
5.66250275226056, 111.85389541494, 18.1995988898424, 79.5127602899565
), SBP_23h_Late.Mid.Rings_2 = c(27.6455730547316, 2.04450518969984, 
3.11285673303378, 173.335869764135, 3.75799617308546, 111.076770881732
), SBP_08h_Early.Trophs_3 = c(63.9441758034038, 8.783379046428, 
1.36572101399292, 296.461875716015, 2.58960184385856, 183.132522675328
), SBP_11h_Mid.Trophs_3 = c(70.3465549733047, 11.856861649251, 
1.25726103732839, 357.299957208987, 2.81495469320334, 192.931412027969
), SBP_14h_Late.Trophs_3 = c(24.5234329455145, 29.1881967487647, 
7.8567780739673, 1772.21118398556, 22.3035208013784, 168.207454664877
), SBP_17h_Early.Rings_3 = c(15.8780527250301, 21.5556594570105, 
7.02241483912262, 644.958506567788, 190.906286209725, 213.781884925216
), SBP_20h_Early.Mid.Rings_3 = c(5.4041703258959, 3.85169957772944, 
5.4692409481873, 114.947181076967, 19.3388591170259, 85.532444920705
), SBP_23h_Late.Mid.Rings_3 = c(20.6721444393214, 2.08063791434729, 
6.25771469400429, 131.683650813693, 7.05321807501424, 100.520179346894
), SBP_02h_Late.Rings_3 = c(31.6880147939969, 3.29026677329356, 
4.84759449553743, 215.013874125238, 3.49951253434249, 134.56799075684
)), row.names = 3731:3736, class = "data.frame")

all.equal(chab_tpm.df[3731:3736,colnames(check_tpm)],check_tpm)
```

```{r super-pseudogene}

chab_tpm_melt.df <- reshape2::melt(chab_tpm.df,
                                   variable.name = 'sample',
                                   value.name = 'tpm') %>% 
  mutate(time = str_extract(sample, 
                            pattern = '([[:digit:]]{2}h|sch.culture)'),
         transmission = str_extract(sample,
                                    pattern = 'rMT|SBP'))

(plot <- ggplot(data = filter(chab_tpm_melt.df, 
                     Geneid %in% 'PCHAS_0420800'), 
       aes(x = time, 
           y = tpm,
           fill = transmission)) +
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "bar", width = 0.75, size = 0.25, position = 'dodge') +
  geom_point(stat = 'identity', position = position_dodge(width = 0.75)) +
  scale_fill_manual(values = cbPalette) +
  theme_classic() +
  theme(axis.text.x=element_text(angle = 90, vjust = 0.5)))

ggsave(filename = 'plots/pseudogene_tpm_barchart.png',
       plot)
  
```


```{r tpm_rescale}

#Setting chab_tpm_ct core dataset - remove pseudogene and re-scaled

chab_tpm_nopseudo <- chab_tpm.df %>% 
  filter(!Geneid %in% 'PCHAS_0420800')

chab_tpm_rescale_nopseudo <- lapply(colnames(chab_tpm_nopseudo), 
                                          function(column_name){
                                            column <- getElement(chab_tpm_nopseudo, 
                                                                 paste(column_name))
                                            if(is.numeric(column)){
                                              #is.numeric to avoid trying to scale the Geneid column
                                              tpm_scale_factor <- 1e6 / sum(column)
                                              tpm_scaled <- as.numeric(column*tpm_scale_factor)
                                            }else{
                                              tpm_scaled <- column
                                            }
                                            return(tpm_scaled)
                                          }) %>% 
  bind_cols

colnames(chab_tpm_rescale_nopseudo) <- colnames(chab_tpm_nopseudo)
# apply(chab_tpm_rescale_nopseudo[,-1], 2, sum)

#Setting the re-scaled version as the default from now on
#Setting chab_tpm_ct core dataset - removed pseudogene and re-scaled
# chab_tpm.df <- chab_tpm_rescale_nopseudo
```

```{r clean_up_and_average}
chab_tpm_avg_nopseudo.df <- reshape2::melt(chab_tpm_rescale_nopseudo, 
                           variable.name = 'sample',
                           value.name = 'tpm') %>% 
  mutate(stage = str_extract(sample, pattern = '.+(?=_[[:digit:]])')) %>% 
  group_by(Geneid, stage) %>% 
  summarise(mean_tpm = mean(tpm)) %>% 
  reshape2::dcast(Geneid ~ stage)
```

```{r save}
write_csv(chab_tpm.df,
          file = 'data/Pchabaudi_TPM_all_withPseudo.csv')

write_csv(chab_tpm_rescale_nopseudo,
          file = 'data/Pchabaudi_TPM_all.csv')

write_csv(chab_tpm_avg_nopseudo.df,
          file = 'data/Pchabaudi_TPM_average.csv')
```