---
title: "Manuscript_Github_Pchabaudi_compBerghei"
author: "Timothy Little"
date: "29/07/2022"
output: html_document
---

Comparing the P. chabaudi and P. berghei pir expression patterns for any similarities.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      results = TRUE, 
                      message = FALSE, 
                      warning = FALSE)

knitr::opts_knit$set(root.dir = 
                       ".")
#Note that the above doesn't give an error if the dir doesn't exist

options(scipen = 999) # Means that numbers are displayed normally not as 1e6 (1000000)
```

```{r loading_packages, include = FALSE, message = FALSE}
# Load the required packages

library('readr')
library('dplyr')
library('ggplot2')
library("RColorBrewer")
library('data.table')
library('tibble')
library('stringr')
library('kableExtra')
library('viridis')
library('reshape2')
library('ComplexHeatmap')
library('circlize') # for circular plots and colour
library("factoextra")
library("MatrixGenerics") # for rowMins and rowMax

#Colour blind friendly palette
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

#Want the defaults for arrange, mutate and summarise to be from dplyr
arrange <- dplyr::arrange
mutate <- dplyr::mutate
summarise <- dplyr::summarise
select <- dplyr::select

#Global save var - set to TRUE if you want to save figures.
global_save_var <- FALSE
```

```{r sourcing_functions}
source('../rfunctions/uniquify_fun.R')
source('../rfunctions/ordering_fun.R')
source('../rfunctions/tpm_fun.R')
source('../rfunctions/usefulFunctions_mround_my-max_firstup_fun.R')
source('../rfunctions/draw_heatmap_zscore.R')
```

```{r ordering_vec}
order_vector <- c("Early.Rings",
                  "Early.Mid.Rings",
                  "Late.Mid.Rings",
                  "Late.Rings",
                  "Rings.Troph.conversion",
                  "Early.Trophs",
                  "Mid.Trophs",
                  "Late.Trophs")
```

```{r cir_id, message=FALSE, results='hide'}
# The _cir_ gene names were downloaded using a search of PlasmoDB _P. chabaudi_ genome. _cir_ genes are the names of _pir_ genes in _P chabaudi_.
# Getting the cir gene names and transcript lengths needed for calculating the transcript-per-million

chabaudi_gene_info <- read.csv('../standard_files/PlasmoDB-47_beta_Pchabaudichabaudi_GenesByTaxon_Summary.csv',
                              header = TRUE,
                              stringsAsFactors = FALSE)

chabaudi_transcript_lengths <- select(chabaudi_gene_info, c(Gene.ID, Transcript.Length)) %>% 
  arrange(Gene.ID)

chabaudi_gene_info <- dplyr::rename(chabaudi_gene_info, 
                             Geneid = Gene.ID,
                             Description = Product.Description)

```

```{r cir_info}

cir_info <- read_delim('../standard_files/Pchabaudi_pir_info.txt',
           delim = '\t') %>% 
  group_by(subfam) %>% 
  mutate(subfam_total_members = n_distinct(Geneid)) %>% 
  mutate(
    subfam_num = paste0(
      subfam, 
      ' (n = ',
      subfam_total_members,
      ')'
    )
  )
cir_id <- cir_info$Geneid

cir_clade_info <- read_delim('../standard_files/Pchabaudi_pir_clades.txt',
            delim = '\t')
```

```{r times_stages_labels}
#Use the RMT stage differential count in order to name the samples, but keep time to retain order
times_stages_table <- data.frame(time = c('02h', '05h', '08h', '11h', '14h', '17h','20h','23h'),
                                 stage_diff_count = c('Late-Rings',
                                                      'Rings-Troph-conversion',
                                                      'Early-Trophs',
                                                      'Mid-Trophs',
                                                      'Late-Trophs',
                                                      'Early-Rings',
                                                      'Early-Mid-Rings',
                                                      'Late-Mid-Rings'))
# R will convert the dashes to '.' at some point so let's have this notation here also
times_stages_table$r_stage_diff_count <- str_replace_all(times_stages_table$stage_diff_count, 
                                                         pattern = '-', 
                                                         replacement = '\\.')
#Add a number from 1 to 8 for each stage, useful for phase calculations
times_stages_table$num <- 1:8
```

```{r colours_clades_phase}
SorL_col <- setNames(
  c(    '#F0E442',
        '#56B4E9', #L
        '#E69F00', #S
        '#999999'),
  sort(
    unique(cir_info$SorL)
  )
)

SorL_col_num <- SorL_col
names(SorL_col_num) <- sapply(names(SorL_col), 
                              function(clade_name) {
                                paste0(
                                  clade_name, 
                                  ' (n = ',
                                  sum(cir_info$SorL == clade_name),
                                  ')'
                                )
                              })

clade_col <- setNames(
  c('#F0E442',
    '#56B4E9', #L1
    '#009E73',
    '#0072B2',
    '#E69F00', #S1
    '#D55E00',
    '#CC79A7',
    'black'),
  sort(
    unique(cir_info$subfam)
  )
)

clade_col_num <- clade_col
names(clade_col_num) <- sapply(names(clade_col), 
                               function(clade_name) {
                                 paste0(
                                   clade_name, 
                                   ' (n = ',
                                   sum(cir_info$subfam == clade_name),
                                   ')'
                                 )
                               })

chapl_aapl_name_col <- setNames(
  c('#F0E442',
    '#56B4E9', #L1
    '#009E73',
    '#0072B2',
    '#E69F00', #S1
    '#D55E00',
    '#CC79A7',
    '#999999'),
  sort(
    unique(cir_info$chapl_aapl_name)
  )
)


chapl_aapl_name_col_num <- chapl_aapl_name_col
names(chapl_aapl_name_col_num) <- sapply(
  names(chapl_aapl_name_col), 
  function(clade_name) {
    paste0(
      clade_name, 
      ' (n = ',
      sum(cir_info$chapl_aapl_name == clade_name, 
          na.rm = TRUE),
      ')'
    )
  })
```

```{r read_TPM}
#Need to use the calculateTPM script first.

chab_tpm.df <- read_csv('data/Pchabaudi_TPM_all.csv')

chab_tpm_avg.df <- read_csv('data/Pchabaudi_TPM_average.csv')

# chab_tpm_avg.list <- list(
#   'RMT' = list(
#     tpm = chab_tpm_avg.df[,grepl(colnames(chab_tpm_avg.df), 
#                                 pattern = 'rMT|Geneid')],
#     experiment = 'RMT'
#   ),
#   'SBP' = list(
#     tpm = chab_tpm_avg.df[,grepl(colnames(chab_tpm_avg.df), 
#                                 pattern = 'SBP|Geneid')],
#     experiment = 'SBP'
#   )
# )

```

```{r berghei_info}
bir_clade <- read_delim('../standard_files/Pberghei_pir_clades.txt', 
                        delim = '\t') %>% 
  group_by(clade) %>% 
  mutate(subfam_total_members = n_distinct(Geneid),
         subfam_num = paste0(clade, ' (n = ', n_distinct(Geneid), ')'))
bir_id <- bir_clade$Geneid

bir_clade_col <- setNames(
  c('#F0E442',
    '#56B4E9',
    '#009E73',
    '#00517f',
    '#E69F00',
    '#893c00',
    '#c879cc',
    '#999999',
    'black',
    'red'),
  sort(
   unique(bir_clade$clade)
  )
)

bir_clade_col_num <- bir_clade_col
names(bir_clade_col_num) <- sapply(names(bir_clade_col), 
                               function(clade_name) {
                                 paste0(
                                   clade_name, 
                                   ' (n = ',
                                   sum(bir_clade$clade == clade_name),
                                   ')'
                                 )
                               })
```

```{r berg_tpm}
berghei_tpm <- read.csv('../chapter3_results_bergheiPir/berghei_tpm.csv', 
                        row.names = 1)
```


```{r chab_subfam_normByGeneNum}
chab_tpm_subfam.df.long <- chab_tpm.df %>% 
  # rownames_to_column('Geneid') %>% 
  filter(Geneid %in% cir_id) %>% 
  melt(id.vars = 'Geneid', variable.name = 'sample', value.name = 'tpm') %>% 
  mutate(subfam = cir_info$subfam_num[match(Geneid, cir_info$Geneid)],
         SorL = cir_info$SorL[match(Geneid, cir_info$Geneid)],
         transmission = str_extract(sample, pattern = 'rMT|SBP'),
         time = str_extract(sample, '[[:digit:]]{2}h|sch.culture'),
         name = paste0(time, str_extract(sample, 
                                         pattern = '(_[[:digit:]]$)'))) %>% 
  group_by(subfam, SorL, transmission, time, name) %>% summarise(tpm = sum(tpm))

( plot <- ggplot(filter(chab_tpm_subfam.df.long, 
                        SorL != 'ancestral'), 
                 aes(x = factor(name,
                                levels = ordering(name, 
                                                  order_vector = c(times_stages_table$time, 'sch.culture'))),
                     y = tpm,
                     fill = subfam)) +
    geom_col(position = "fill") +
    xlab('Stages') + ylab('TPM') +
    scale_fill_manual(values = clade_col_num[!grepl(names(clade_col_num), pattern =  'ancestral')]) +
    facet_grid(~transmission, scales = "free_x") +
    # facet_grid(transmission ~ SorL, scales = "free")+
    theme_classic()+
    theme(axis.text.x = element_text(angle = 90,
                                     hjust = 1),
          strip.text = element_text(size = 12),
          axis.text.y = element_text(size = 10),
          axis.title = element_text(size = 12))  +
  guides(fill = guide_legend(title = 'Sub-family')) )

if(global_save_var){
  ggsave(plot,
         file = 'plots/chab_subfamProp.png',
         width = 7,
         height = 8,
         units = "in")
}
```

Dividing the TPM by the number of genes in each sub-family to see TPM/gene.

```{r normalised_subfam}

subfam_number.table <- group_by(cir_info, subfam_num) %>% 
  summarise(subfam_total_members = n_distinct(Geneid))

#Don't use n_distinct(Geneid) to nornalise by number of genes in a subfam group
# There has been some filtering so not all members are in the dataset.
chab_tpm_subfam_norm.df.long <- chab_tpm_subfam.df.long %>% 
  group_by(transmission, time, subfam) %>% 
         mutate(norm_tpm = tpm / cir_info$subfam_total_members[match(Geneid, cir_info$Geneid)])

ggplot(filter(chab_tpm_subfam_norm.df.long, 
              !subfam == 'ancestral (n = 1)'), 
       aes(x = factor(name,
                      levels = ordering(name, 
                                        order_vector = c(times_stages_table$time, 'sch.culture'))),
           y = norm_tpm,
           fill = subfam)) +
  geom_col(position = "fill") +
    xlab('stages') +
  ylab('TPM per gene') +
  facet_wrap(~transmission, scales = "free_x") +
  scale_fill_manual(values = clade_col_num[!names(clade_col_num) == 'ancestral (n = 1)']) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90))

```

```{r berghei_subfam}
berg_tpm_subfam.df.long <- berghei_tpm[,
                                       c('Geneid',
                                         str_subset(colnames(berghei_tpm), 
                                                    pattern = 'Asex\\.(?!Mixed)'))] %>% 
  filter(Geneid %in% bir_id) %>% 
  melt(id.vars = 'Geneid', variable.name = 'sample', value.name = 'tpm') %>% 
  mutate(subfam = bir_clade$subfam_num[match(Geneid, bir_clade$Geneid)],
         experiment = str_extract(sample, pattern = '^E[[:digit:]]{2}'),
         stage = str_extract(sample, 'Asex.+(?=_[[:digit:]])'),
         name = str_extract(sample, '(?<=Asex\\.).+')) %>% 
  group_by(subfam, experiment, stage, name) %>% summarise(tpm = sum(tpm))

( plot <- ggplot(filter(berg_tpm_subfam.df.long, subfam != 'ancestral (n = 1)'), 
                 aes(x = factor(name,
                                levels = ordering(name,
                                                  order_vector = c('Ring','Troph','Schiz','Mixed'))),
                     y = tpm,
                     fill = subfam)) +
    geom_col(position = "fill") +
    ylab("TPM") +
    xlab('Stages') +
    scale_fill_manual(values = bir_clade_col_num[!grepl(names(bir_clade_col_num), 
                                                        pattern =  'ancestral')]) +
    facet_wrap(~experiment, scales = "free_x")+
    theme_classic()+
    theme(axis.text.x = element_text(angle = 90,
                                     hjust = 1),
          strip.text = element_text(size = 12),
          axis.text.y = element_text(size = 10),
          axis.title = element_text(size = 12))
)

if(global_save_var){
  ggsave(plot,
         file = 'plots/berghei_subfamProp_AsexExact.png',
         width = 7,
         height = 8,
         units = "in")
}
```

```{r berghei_subfam_normByGeneNum}

#Don't use n_distinct(Geneid) to nornalise by number of genes in a subfam group
# There has been some filtering so not all members are in the dataset.
berg_tpm_subfam_norm.df.long <- berg_tpm_subfam.df.long %>% 
  group_by(experiment, stage, subfam) %>% 
  mutate(norm_tpm = tpm / bir_clade$subfam_total_members[match(Geneid, bir_clade$Geneid)])

ggplot(filter(berg_tpm_subfam_norm.df.long, 
              !subfam == 'ancestral (n = 1)'), 
       aes(x = factor(name,
                      levels = ordering(name, 
                                        order_vector = c('Ring','Troph','Schiz','Mixed'))),
           y = norm_tpm,
           fill = subfam)) +
  geom_col(position = "fill") +
  xlab('stages') +
  ylab('TPM per gene') +
  facet_wrap(~experiment, scales = "free_x") +
  scale_fill_manual(values = bir_clade_col_num[!names(bir_clade_col_num) %in% c('ancestral (n = 1)',
                                                                                'S6 (n = 1)')]) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90))
```

Hypothesis: There is a difference in the transcription level of L-form pirs as a proportion of all S/L-form pirs between Asexual Schizont samples and non-schizont samples

```{r chab_prop_test}
chab_tpm_subfam.L.df.long <- chab_tpm_subfam.df.long %>% 
  filter(!subfam == 'ancestral (n = 1)') %>% 
  mutate(subfam_L = grepl(subfam, pattern = '^L')) %>% 
  group_by(transmission, time, name, sample, subfam_L) %>% 
  summarise(sum_tpm = sum(tpm)) %>% 
  group_by(sample) %>% 
  mutate(sum_tpm_all = sum(sum_tpm)) %>% 
  filter(subfam_L)

chab_tpm_subfam.L.df.long %>% 
  group_by(sample) %>% 
  mutate(prop_L = sum_tpm/sum_tpm_all) %>% 
  arrange(desc(prop_L))

bind_rows(
  lapply(str_subset(unique(chab_tpm_subfam.L.df.long$sample), 
                    pattern = 'sch.culture',
                    negate = TRUE),
         function(sample.name){
           bind_rows(
             lapply(c('rMT_sch.culture_1', 
                      'rMT_sch.culture_2', 
                      'rMT_sch.culture_3'), 
                    function(sch.cult){
                      L_forms <- c(chab_tpm_subfam.L.df.long$sum_tpm[
                        chab_tpm_subfam.L.df.long$sample == sch.cult 
                      ],
                      chab_tpm_subfam.L.df.long$sum_tpm[
                        chab_tpm_subfam.L.df.long$sample == sample.name
                      ])
                      SorL_forms <- c(chab_tpm_subfam.L.df.long$sum_tpm_all[
                        chab_tpm_subfam.L.df.long$sample == sch.cult 
                      ],
                      chab_tpm_subfam.L.df.long$sum_tpm_all[
                        chab_tpm_subfam.L.df.long$sample == sample.name
                      ])
                      test.stat <- prop.test(L_forms, SorL_forms,
                                             alternative = 'greater')
                      data.frame(comp = paste0(sample.name, '-', sch.cult),
                                 pval = test.stat$p.value)
                    })
           )
         })
)%>% 
  mutate(padj = round(p.adjust(pval), digits = 5),
         pval = round(pval, digits = 5)) %>% 
  arrange(desc(padj))
```

```{r berg_prop_test}
berg_tpm_subfam.L.df.long <- berg_tpm_subfam.df.long %>% 
  filter(!subfam == 'ancestral (n = 1)') %>% 
  mutate(subfam_L = grepl(subfam, pattern = '^L')) %>% 
  group_by(experiment, name, sample, subfam_L) %>% 
  summarise(sum_tpm = sum(tpm)) %>% 
  group_by(sample) %>% 
  mutate(sum_tpm_all = sum(sum_tpm)) %>% 
  filter(subfam_L)

berg_tpm_subfam.L.df.long %>% 
  group_by(sample) %>% 
  mutate(prop_L = sum_tpm/sum_tpm_all) %>% 
  arrange(desc(prop_L))

bind_rows(
  lapply(str_subset(unique(berg_tpm_subfam.L.df.long$sample), 
                    pattern = 'Schiz|Mixed',
                    negate = TRUE),
         function(sample.name){
           bind_rows(
             lapply(str_subset(unique(berg_tpm_subfam.L.df.long$sample), 
                               pattern = 'Schiz'), 
                    function(sch.cult){
                      L_forms <- c(berg_tpm_subfam.L.df.long$sum_tpm[
                        berg_tpm_subfam.L.df.long$sample == sch.cult 
                      ],
                      berg_tpm_subfam.L.df.long$sum_tpm[
                        berg_tpm_subfam.L.df.long$sample == sample.name
                      ])
                      SorL_forms <- c(berg_tpm_subfam.L.df.long$sum_tpm_all[
                        berg_tpm_subfam.L.df.long$sample == sch.cult 
                      ],
                      berg_tpm_subfam.L.df.long$sum_tpm_all[
                        berg_tpm_subfam.L.df.long$sample == sample.name
                      ])
                      test.stat <- prop.test(L_forms, SorL_forms,
                                             alternative = 'greater')
                      data.frame(comp = paste0(sample.name, '-', sch.cult),
                                 pval = test.stat$p.value)
                    })
           )
         })
)%>% 
  mutate(padj = round(p.adjust(pval), digits = 5),
         pval = round(pval, digits = 5)) %>% 
  arrange(desc(padj))
```

