---
title: "P. berghei pir - Sample clustering/PCA"
output: 
  html_document:
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

```{r setup, include=FALSE, message = FALSE}
knitr::opts_chunk$set(echo = FALSE, results = TRUE, message = FALSE, warning = FALSE)
options(scipen = 999) # Means that numbers are displayed normally not as 1e6 (1000000)
```

```{r load_packages, include = FALSE, message = FALSE}
library('readxl')
library('readr')
library('dplyr')
library('purrr')
library('ggplot2')
library("RColorBrewer")
library('colorspace')
library('data.table')
library('gridExtra')
library('tibble')
library('scales')
library('stringr')
library('kableExtra')
library('viridis')
library('corrplot')
library('reshape2')
library('circlize') # for circular plots
library('cluster')    # clustering algorithms
library('factoextra') # clustering algorithms & visualization
library('ggfortify') # so ggplot understands prcomp objects
library('ape') # for phylogeny plotting
library('dendextend') # for dendrogram plotting
library('ggrepel')

#Colour blind friendly palette
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
show_col(cbPalette)

#Want the defaults for arrange, select, mutate and summarise to be from dplyr
arrange <- dplyr::arrange
mutate <- dplyr::mutate
summarise <- dplyr::summarise
select <- dplyr::select

#Save plots?
global_save_var <- FALSE

```

```{r sourcing_functions}
source('../rfunctions/uniquify_fun.R')
source('../rfunctions/ordering_fun.R')
source('../rfunctions/tpm_fun.R')
source('../rfunctions/usefulFunctions_mround_my-max_firstup_fun.R')
```

```{r pir_id, message=FALSE, results='hide'}

# Getting the pir gene names and transcript lengths needed for calculating the transcript-per-million

# First we need to get the _pir gene names. These have been downloaded using a search of PlasmoDB P. berghei ANKA genome v3 using  "\*pir\*" search and manual removal of genes which clearly were not pir.

allgene.info.file <- '../standard_files/PlasmoDb-47_beta_Pberghei_GenesByTaxon_Summary.csv'
allgene.info <- as_tibble(
  read_csv(allgene.info.file)
) %>%
  mutate(Geneid = `Gene ID`, 
         Description = `Product Description`,
         is_pseudo = `Is Pseudo`,
         transcript_length = `Transcript Length`,
         paralog_count = `Paralog count`) %>% 
  select(Geneid, 
         Description, 
         is_pseudo, 
         transcript_length, 
         paralog_count)

pir_info <- allgene.info %>% 
  filter(str_detect(Description, 
                    pattern = 'PIR(?!.+pseud)'), 
         is_pseudo == 'No') %>% 
  filter(!Geneid %in% 'PBANKA_0524600')
#Removing the pir-like gene.

pir_id <- pir_info$Geneid
```

## Reading the TPM 

```{r read_tpm}
berghei_tpm <- read.csv('berghei_tpm.csv', 
                        row.names = 1)
```

## Density plots

```{r bandwidth}
# Compute the average bandwidth for all the samples so that we have the same value for each figure.
list_bw = lapply(expt_tpms, function(expt){ 
  lapply(colnames(expt$tpm)[-1], 
         function(sample){
           data = as.numeric(unlist(select(expt$tpm, 
                                           paste(sample))))
           dens = density(log1p(data))
           dens$bw
         })
})
bw_mean = mean(unlist(list_bw))
bw_range = max(unlist(unlist(list_bw))) - min(unlist(unlist(list_bw)))
print(paste('Bandwidth range = ', bw_range))
```

```{r density_plots}
source('../rfunctions/density_samples_fun.R')
lapply(expt_tpms, 
       density_samples, 
       save = global_save_var)
```

```{r melt_df}
berghei_tpm_melt <- berghei_tpm %>% 
    #Melt the data frame for using dplyr
    melt(id.vars = 'Geneid', variable.name = 'sample', value.name = 'tpm') %>% 
    mutate(experiment = unlist(lapply(strsplit(as.character(sample), split = '_'),`[`,1)),
           stage = unlist(lapply(strsplit(as.character(sample), split = '_'),`[`,2)))

berghei_tpm_avg_byexpt <- berghei_tpm_melt %>% 
  group_by(Geneid, experiment, stage) %>%  
  dplyr::summarise(median_tpm = median(tpm),
                   mean_tpm = mean(tpm), 
                   min_tpm = min(tpm), 
                   max_tpm = max(tpm), 
                   sd_tpm = sd(tpm)) %>% 
  mutate(expt_stage = paste0(experiment, '_', stage))
```

## The highest expressed _pir_ genes across the samples

```{r top_genes, echo = FALSE, message = FALSE, results = "asis"}
top_n_genes <- function(tpm_avg, 
                        n = 5, 
                        threshold = 0){
  #Calculating the index used to make the kable table
  
  kable_index = rep(n, times = length(unique(tpm_avg$stage)))
  names(kable_index) = unique(tpm_avg$stage)
  
  tpm_avg <- tpm_avg %>% 
    ungroup %>% 
    group_by(stage)
  
  all_genes = tpm_avg %>% 
    slice_max(order_by = mean_tpm, n = n) %>% 
    dplyr::arrange(stage, desc(mean_tpm)) %>% 
    mutate_if(is.numeric, function(x){round(x, digits = 2)}) %>% 
    mutate(mean_tpm = cell_spec(mean_tpm, "html", 
                                color = ifelse(mean_tpm > threshold, "red", "blue")))
  all_genes$short_descr = allgene.info$handy_descriptions[match(all_genes$Geneid, 
                                                                allgene.info$Geneid)]
  all_genes$description = allgene.info$Description[match(all_genes$Geneid, 
                                                         allgene.info$Geneid)]
  all_genes = all_genes %>% 
    kable(format = 'html', 
          caption = paste0(tpm_avg$experiment[1], ' - Top ', n, ' expressed total genes'), 
          escape = FALSE) %>%
    kable_styling() %>% 
    pack_rows(index = kable_index)
  
  pir_genes = tpm_avg %>% 
    filter(Geneid %in% pir_id) %>% 
    slice_max(order_by = mean_tpm, n = n) %>% 
    dplyr::arrange(stage, desc(mean_tpm)) %>% 
    mutate_if(is.numeric, function(x){round(x, digits = 2)}) %>% 
    mutate(mean_tpm = cell_spec(mean_tpm, "html", 
                                color = ifelse(mean_tpm > threshold, "red", "blue"))) %>% 
    mutate(Geneid = as.character(Geneid)) %>% 
    mutate(Description = allgene.info$handy_descriptions[match(Geneid, 
                                                               allgene.info$Geneid)]) %>% 
    mutate(Geneid = ifelse(grepl(Geneid,
                                 pattern = 'PBANKA_0100500'),
                           cell_spec(Geneid, bold = TRUE),
                           Geneid)) %>%
    kable(format = 'html', 
          caption = paste0(tpm_avg$experiment[1], ' - Top ', n, ' expressed pir genes'), 
          escape = FALSE) %>% 
    pack_rows(index = kable_index) %>% 
    kable_styling()
  
  # print(paste0('Completed analysis of ', tpm_input$experiment))
  list = list(all_genes, pir_genes, unique(tpm_avg$experiment))
  names(list) = c('all_genes', 'pir_genes', 'experiment')
  return(list)
}

### Generate and save top genes - core dataset all_top

all_top <- lapply(unique(berghei_tpm_avg_byexpt$experiment), 
                  function(expt) top_n_genes(berghei_tpm_avg_byexpt[berghei_tpm_avg_byexpt$experiment == expt,]))


```

##Pearson correlation plot of collated samples.

```{r all_samples_collated_heatmap}

## Collated across the stages for a more easily understood heatmap
big_top <- berghei_tpm_avg_byexpt

big_top_recast = dcast(big_top, Geneid ~ expt_stage, value.var = 'mean_tpm')
big_top_recast <- column_to_rownames(big_top_recast, var = 'Geneid')
# big_top_recast = big_top_recast[,-1,drop=FALSE]

# apply(big_top_recast, 2, sum)

if(global_save_var){
  png(file = 'plots/thesis_chapter3_figure_corrplot.png',
      height = 5,
      width = 5,
      units = "in",
      res = 330)
}
corrplot(cor(big_top_recast), method = 'color', 
         order = 'hclust', hclust.method = 'ward.D2',
         cl.lim = c(-0.036,1), tl.cex = 0.50,
         col = c(viridis(100), magma(115)), addrect = 6)
if(global_save_var){
  dev.off()
}
```

## Supp figure 1B - PCA of all the samples.

```{r supp_figure_PCA_allgenes}
#Set the Geneids as the berghei_tpm rownames so that the PCA function retains them as loading variables
big_tpm_pca <-  prcomp(column_to_rownames(berghei_tpm, var = 'Geneid'), 
                       center = TRUE, 
                       scale. = TRUE)

big_tpm_pca_var <- get_pca_var(big_tpm_pca)
big_tpm_pca_coords <- data.frame(sample = rownames(big_tpm_pca_var$coord),
                                 big_tpm_pca_var$coord)
big_tpm_pca_coords$stage_experiment <- str_extract(
  big_tpm_pca_coords$sample,
  pattern = '.+(?=_[[:digit:]]+$)'
)
big_tpm_pca_coords$stage <- str_extract(big_tpm_pca_coords$sample, 
                                        pattern = '(?<=E[[:digit:]]{2}_).+(?=_[[:digit:]]+$)')
big_tpm_pca_coords$experiment <- str_extract(big_tpm_pca_coords$sample, 
                                             pattern = '^.+(?=_.+_[[:digit:]]+$)')

colornames <- setNames(rep_len(c(cbPalette, 'black', 'red'), 
                               length.out = length(unique(big_tpm_pca_coords$stage))),
                       unique(big_tpm_pca_coords$stage))

pca_importance <- as.data.frame(summary(big_tpm_pca)$importance)

(
  plot <- ggplot(big_tpm_pca_coords,
       aes(
         x = Dim.1,
         y = Dim.2,
         col = stage,
         shape = experiment
       )) +
  geom_point() +
  scale_fill_manual(values = colornames) +
  scale_shape_manual(
    values = setNames(
      c(1:10),
      unique(big_tpm_pca_coords$experiment)
    )
  ) +
  xlab(paste0('PC1 ', round(pca_importance$PC1[2]*100, digits = 1),'%')) +
  ylab(paste0('PC2 ', round(pca_importance$PC2[2]*100, digits = 1),'%')) +
  guides(shape=guide_legend(ncol=2),
         col = guide_legend(ncol=2)) +
  theme_classic() +
  theme(legend.key.size = unit(5, 'mm'),
        legend.text = element_text(size = 6))
)

if(global_save_var){
  ggsave('plots/thesis_chapter3_figure_pca_allGenes.png',
         width = 7,
         height = 5,
         units = "in",
         plot)
}

```

# Biplot of the PCA using all genes.

```{r biplot}
set.seed(982)

pca_object_loadings <- data.frame(big_tpm_pca$x)

#Remove the large PCA object
rm(big_tpm_pca)

#Calculate the highest loadings, only want these labels
top_loadings <- 15
pca_object_loadings$Geneid <- ''

highest_loadings <- pca_object_loadings[rev(
    order(
      apply(
        abs(
          data.frame(pca_object_loadings$PC1, 
                     pca_object_loadings$PC2)),
        1,
        sum)
    )
)[1:top_loadings],]

highest_loadings$Geneid <- 
  gsub(
    gsub(allgene.info$Description[match(rownames(highest_loadings),
                                      allgene.info$Geneid)], 
       pattern = 'putative',
       replacement = 'put.'),
    pattern = 'early transcribed membrane protein',
    replacement = 'ETRAMP')

loadings_scale <- 0.005

(bp <- plot +
    geom_point(data = highest_loadings,
               aes(x = PC1*loadings_scale, y = PC2*loadings_scale),
               inherit.aes = FALSE, size = 0.75) +
    geom_text_repel(data = highest_loadings,
                    aes(x = PC1*loadings_scale, y = PC2*loadings_scale,
                        label = Geneid),
                    inherit.aes = FALSE, 
                    size = 3,
                    min.segment.length = 0,
                    force = 100
    ) )

if(global_save_var){
  ggsave('plots/thesis_chapter3_figure_biplot.png',
         width = 7,
         height = 5,
         units = "in",
         bp)
}

#Note that CSP is the 140 highest laodings gene.
```

#Supp figure 1C - PCA using just the pir genes.

```{r supp_figure_PCA_pir}
big_tpm_pir <- berghei_tpm[berghei_tpm$Geneid %in% pir_id,-1,drop = FALSE]

big_tpm_pir <- big_tpm_pir[apply(big_tpm_pir, 2, sum) >= 50]

big_tpm_pir_pca <-  prcomp(big_tpm_pir, 
                       center = TRUE, 
                       scale. = TRUE)

big_tpm_pir_pca_var <- get_pca_var(big_tpm_pir_pca)
big_tpm_pir_pca_coords <- data.frame(sample = rownames(big_tpm_pir_pca_var$coord),
                                 big_tpm_pir_pca_var$coord)
big_tpm_pir_pca_coords$stage_experiment <- str_extract(
  big_tpm_pir_pca_coords$sample,
  pattern = '.+(?=_[[:digit:]]+$)'
)
big_tpm_pir_pca_coords$stage <- str_extract(big_tpm_pir_pca_coords$sample, 
                                        pattern = '(?<=E[[:digit:]]{2}_).+(?=_[[:digit:]]+$)')
big_tpm_pir_pca_coords$experiment <- str_extract(big_tpm_pir_pca_coords$sample, 
                                             pattern = '^.+(?=_.+_[[:digit:]]+$)')

colornames <- setNames(rep_len(c(cbPalette, 'black', 'red'), 
                               length.out = length(unique(big_tpm_pir_pca_coords$stage))),
                       unique(big_tpm_pir_pca_coords$stage))

pca_importance <- as.data.frame(summary(big_tpm_pir_pca)$importance)

(
  plot_pir <- ggplot(big_tpm_pir_pca_coords,
                     aes(
                       x = Dim.1,
                       y = Dim.2,
                       col = stage,
                       shape = experiment
                     )) +
    geom_point() +
    scale_fill_manual(values = colornames) +
    scale_shape_manual(
      values = setNames(
        c(1:10),
        unique(big_tpm_pca_coords$experiment)
      )
    ) +
    xlab(paste0('PC1 ', round(pca_importance$PC1[2]*100, digits = 1),'%')) +
    ylab(paste0('PC2 ', round(pca_importance$PC2[2]*100, digits = 1),'%')) +
    guides(shape=guide_legend(ncol=2),
           col = guide_legend(ncol=2)) +
    theme_classic() +
    theme(legend.key.size = unit(5, 'mm'),
          legend.text = element_text(size = 6)) 
)

if(global_save_var){
    ggsave('plots/thesis_chapter3_figure_PCA_pir.png',
         width = 7,
         height = 5,
         units = "in",
         plot_pir)
}
```