---
title: "tl22-01_VvxSalmonAlignment"
output: 
  html_document:
    theme: cosmo
    code_download: true
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

# Intro

This script analyses the results of a _de novo_ assembly of Plasmodium vivax RNAseq data.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      results = TRUE, 
                      message = FALSE, 
                      warning = TRUE)
options(scipen = 999) 
# Means that numbers are displayed normally not as 1e6 (1000000)
```

```{r loading_packages, include = FALSE, message = FALSE}
# Load the required packages
library('readr')
library('dplyr', 
        lib.loc = "/camp/home/littlet/R/x86_64-pc-linux-gnu-library/4.0")
library('stringr')
library('purrr')
library('kableExtra')
library('plotly')
library('rmarkdown')
library('seqinr')
```

```{r function_ordering}
ordering <- function(to_order, order_vector, fromLast = TRUE){
  order = unlist(sapply(order_vector, 
                        function(x){unique(str_subset(to_order, 
                                                      pattern = paste(x)))}))
  order_missing = unique(to_order[!to_order %in% order])
  unique(c(order, order_missing), fromLast = fromLast)
}
```

```{r function_uniquify}
#Make unique identifiers from a vector of names by appending '_X' where X is the number of the 'replicate'.
uniquify <- function(vector_names, first_one = TRUE){
  new_vector <- c()
  df_num <- data.frame(names = unique(vector_names), rep = 1)
  if(first_one){
    duplicated_names <- unique(vector_names[duplicated(vector_names)])
  }else{
    duplicated_names <- c()
  }
  for(i in 1:length(vector_names)){
    if(vector_names[i] %in% duplicated_names|!first_one){
      name <- paste(vector_names[i],
                    df_num$rep[match(vector_names[i], df_num$names)],
                    sep = '_')
      #Increase rep counter
      df_num$rep[match(vector_names[i], df_num$names)] <- df_num$rep[match(vector_names[i], df_num$names)] + 1
      new_vector[i] <- name
    }else{
      new_vector[i] <- vector_names[i]
    }
  }
  new_vector
}
```

```{r load_data}
pir.info <- read.csv('/camp/lab/langhornej/working/HPC/littlet/tl22-01_vivaxDeNovo/github_project/tl22-01_vivaxDeNovo/PlasmoDBv56_VvxPirs_P01_Sal1.csv', 
                     header = TRUE)
```

```{r de_novo_pir_list}
# pir_lists.dir <- "/camp/lab/langhornej/working/HPC/littlet/tl22-01_vivaxDeNovo/hmmer/scspades/"
# pir_lists.files <- list.files(pir_lists.dir, 
#                               pattern = '0.95cdhitest_hmmerPfam-PlasmodiumVir.list',
#                               full.names = TRUE) %>% 
#   #Get those with file sizes above zero (no empty lists)
#   # file.info gets the file sizes
#   .[file.info(., extra_cols = FALSE)$size > 0]
# 
# # Read the pirs into a list (as a vector).
# # sapply through the file list, read in the table and subset with the automatic column name V1
# pir_list.vec <- unlist(
#   sapply(pir_lists.files, 
#          function(pir_list){
#            #str_removing the _1 at the end of the removal 
#            str_remove(read.table(pir_list)$V1, 
#                       pattern =  '_[[:digit:]]$')
#          })
# )

#Need to read in the >200aa ORF length  pirs
pir_list.vec <- str_extract(names(read.fasta('/camp/lab/langhornej/working/HPC/littlet/tl22-01_vivaxDeNovo/hmmer/scspades/G18_B20_C19_Z16_M19_R20_K1719_G19-cat/G18_B20_C19_Z16_M19_R20_K1719_G19-cat_scspades_0.95cdhitest_hmmerPfam-PlasmodiumVir_200aa.fasta')),
                                  pattern = 'NODE.+(?=_[[:digit:]])')
                            
print(paste0('Number of pirs total: ', length(pir_list.vec)))
```

```{r load_cluster_data}
node.df <- read.csv('/camp/lab/langhornej/working/HPC/littlet/tl22-01_vivaxDeNovo/networks/tpmOver1/sequence-MCLclusters_tpmOver1_30pid25lngthI14.csv') %>% 
  #Convert the cluster numbers to a factor as they are discrete categories.
  mutate(across(contains('lngth'), factor))
```

```{r define_subfam}
cluster_subfam.df <- read.csv('/camp/lab/langhornej/working/HPC/littlet/tl22-01_vivaxDeNovo/networks/tpmOver1/cluster_subfam_30pid25lngthI1.4.csv')

cluster_subfam_25.df <- head(cluster_subfam.df, 
                             n = 25)

# write.csv(cluster_subfam_25.df,
#           file = 'cluster_subfam_25_df.csv',
#           row.names = FALSE)

# rowwise makes mutate go row by row instead of vectorising across the whole X30pid25lngthI1.4 column
node_subfam.df <- rowwise(node.df) %>% 
  mutate(subfam_assignment = cluster_subfam.df$subfam[cluster_subfam.df$X30pid25lngthI1.4 %in% X30pid25lngthI1.4],
         pir_label = str_extract(label,
                                 pattern = 'NODE_.+(?=_[[:digit:]])'))

```

unassigned = No Sal1 sequences in the cluster for subfamily assignment.

```{r load_salmon}
#Get the full file paths of the salmon quant.sf files.
quant_files <- list.files(path = "/camp/lab/langhornej/working/HPC/littlet/tl22-01_vivaxDeNovo/salmon/all_files",
                          recursive = TRUE,
                          pattern = 'quant.sf',
                          full.names = TRUE)

#Load the samplesheet.txt file of metadata, add new column for salmon files
samplesheet <- read.delim("/camp/lab/langhornej/working/HPC/littlet/tl22-01_vivaxDeNovo/samplesheet.txt") %>% 
  mutate(salmon_quant = '')

#Make a table of the files and the sample metadata from samplesheet.txt
# for loop over all the quant_files that exist
for(quant_name in quant_files){
  #Looking for the sample prefix in the file path between 'all_files' and '{prefix}_quant'
  prefix <- str_extract(quant_name, 
                        pattern = '(?<=\\/all_files/).+(?=/.+_quant)')
  if(nrow(samplesheet[grep(samplesheet$prefix, 
                   pattern = prefix),]) > 1){
                     stop(paste0(prefix,
                                 " matches more than one row of the samplesheet"))
                   }
  samplesheet[grep(samplesheet$prefix, 
                   pattern = prefix),'salmon_quant'] <- quant_name
}

sum(grepl(samplesheet[,'salmon_quant'], pattern = 'camp')) == 247
```

Make a list element of all the quant values.

```{r list_expt_quant}
expt_quant <- lapply(unique(samplesheet$Experiment),
                     #For each experiment name
                     function(expt){
                       print(paste0('Starting ', expt))
                       #Get the experimental metadata by subsetting
                       expt_data <- samplesheet[samplesheet$Experiment == expt,]
                       if(!sum(grepl(expt_data$assembly, pattern = '.')) == 0){
                         #If sum of grepping for any character is not zero, do the analysis 
                         # i.e there are assembly names and probably assemblies/quant files
                         # For each individual assembly 
                         # (we can bind rows for samples that use the same assembly)
                         assembly_quants <- lapply(unique(expt_data$assembly), 
                                                   function(assembly){
                                                     #Get the prefix(es) that were quantified against this assembly
                                                     prefix <- expt_data[expt_data$assembly == assembly, 
                                                                         'prefix']
                                                     #Get the quant file name(s) and read it in
                                                     quant_file <- expt_data[expt_data$prefix %in% prefix,
                                                                             'salmon_quant']
                                                     
                                                     #If the file name isn't zero length (i.e. there is no name)
                                                     if(!length(quant_file) == 0){
                                                       #If length is one then we can make the quant file for just one sample 
                                                       if(length(prefix) == 1){
                                                         quant <- read.delim(quant_file)
                                                         #Subset the Name and TPM columns and keep it as a data frame (drop = F)
                                                         quant.df <- quant[,c('Name','TPM'), drop = FALSE]
                                                         
                                                         # Give a warning if the TPM values are all zero.
                                                         if(sum(as.double(quant$TPM)) == 0){
                                                           #Do not return a quant.df data.frame, making it nothing:
                                                           quant.df <- ''
                                                           print(paste0(prefix, 
                                                                        " associated file has recorded no TPM values"))
                                                           #empty return to end the lapply for this file
                                                           return()
                                                         }
                                                         
                                                       } else {
                                                         #If longer than 1 so there's more than one quant file for this assembly
                                                         quant.rlist <- lapply(quant_file, read.delim)
                                                         #Use purrr reduce to bind the df in the list together with inner_join,
                                                         # combining by the 'Name' column which should be the same throughout.
                                                         quant.df <- purrr::reduce(quant.rlist, inner_join, by = 'Name')
                                                         quant.df <- quant.df[grepl(colnames(quant.df), pattern = 'Name|TPM')]
                                                         
                                                         # Give a warning if the TPM values are all zero.
                                                         if(any(sapply(quant.df[grepl(colnames(quant.df), 
                                                                                      pattern = 'TPM')], 
                                                                       function(x) sum(as.double(x))) == 0)){
                                                           print(paste0("At least some of ", 
                                                                        expt, 
                                                                        " results have recorded no TPM values"))
                                                           return()
                                                         }
                                                         
                                                       }
                                                       
                                                       #Put the sample name (prefix) in the TPM column
                                                       colnames(quant.df)[grepl(colnames(quant.df), 
                                                                                pattern = 'TPM')] <- paste0(prefix,
                                                                                                            '_TPM')
                                                       
                                                       # Mark the pir NODES as such (TRUE in the 'pir' column)
                                                       quant.df$pir <- quant.df$Name %in% pir_list.vec
                                                       print(paste0('Number of pirs = ', sum(quant.df$pir)))
                                                       
                                                       #Assign the sub-families of the pirs
                                                       # Alter the label of the de novo pirs so they look the same
                                                       pir_labels <- str_extract(node_subfam.df$label,
                                                                                 pattern = 'NODE_.+(?=_[[:digit:]])')
                                                       
                                                       quant.df <- rowwise(quant.df) %>% 
                                                         mutate(subfam = ifelse(Name %in% pir_labels,
                                                                                node_subfam.df$subfam_assignment[
                                                                                  pir_labels %in% Name
                                                                                ], 
                                                                                'noCluster'))
                                                       
                                                       return(quant.df)
                                                     }
                                                   })
                         
                         names(assembly_quants) <- str_extract(unique(expt_data$assembly),
                                                               pattern = '(?<=\\/cdhitest/).+(?=/.+fasta)')
                         print(paste0('Finished ', expt))
                         return(assembly_quants)  
                       }
                     }
)

names(expt_quant) <- unique(samplesheet$Experiment)
```

'noCluster' means that the pir sequence was not in the '/camp/lab/langhornej/working/HPC/littlet/tl22-01_vivaxDeNovo/gephi/G18_B20_C19_Z16_M19_R20_K1719_G19-cat/nodes_mcl_50pid100lngthNoSame2Same.csv' file.
I think this means that they didn't have any BLAST similarity to other pir to the thresholds:
* 50% id.
* alignment >=100aa.

```{r save_quant}
lapply(names(expt_quant), 
       function(expt){
         lapply(names(expt_quant[[expt]]), 
                function(assembly){
                  if(!is.null(expt_quant[[expt]][[assembly]])){
                    #Make the directories to save the files in. Recursive = TRUE so that it will make parents t
                    # dir.create(paste( "/camp/lab/langhornej/working/HPC/littlet/tl22-01_vivaxDeNovo/quantFiles_R",
                    #                   expt, assembly, sep = '/'),
                    #            recursive = TRUE)
                    write.csv(expt_quant[[expt]][[assembly]],
                              file = paste( "/camp/lab/langhornej/working/HPC/littlet/tl22-01_vivaxDeNovo/quantFiles_R",
                                            expt, assembly,
                                            paste0('quant_', 
                                                   assembly,
                                                   '_subfam30pid25lngthI14.csv'),
                                            sep = '/'),
                              row.names = FALSE
                    )
                  }
                })
       })
```

